File: python/schema/dashboards.py
from typing import Any

from pydantic import BaseModel, Field


class DashboardCreatedBy(BaseModel):
    email: str


class Dashboard(BaseModel):
    id: int = Field(gt=0)
    name: str
    description: str | None = None
    pinned: bool | None = None
    created_at: str
    created_by: DashboardCreatedBy | None = None
    is_shared: bool | None = None
    deleted: bool | None = None
    filters: dict[str, Any] | None = None
    variables: dict[str, Any] | None = None
    tags: list[str] | None = None
    tiles: list[dict[str, Any]] | None = None


class CreateDashboardInput(BaseModel):
    name: str = Field(min_length=1, description="Dashboard name is required")
    description: str | None = None
    pinned: bool = False
    tags: list[str] | None = None


class UpdateDashboardInput(BaseModel):
    name: str | None = None
    description: str | None = None
    pinned: bool | None = None
    tags: list[str] | None = None


class ListDashboards(BaseModel):
    limit: int | None = Field(None, gt=0)
    offset: int | None = Field(None, ge=0)
    search: str | None = None
    pinned: bool | None = None


class AddInsightToDashboard(BaseModel):
    insight_id: int = Field(gt=0)
    dashboard_id: int = Field(gt=0)

File: typescript/src/schema/dashboards.ts
import { z } from "zod";

// Base dashboard schema from PostHog API
export const DashboardSchema = z.object({
	id: z.number().int().positive(),
	name: z.string(),
	description: z.string().nullish(),
	pinned: z.boolean().nullish(),
	created_at: z.string(),
	created_by: z
		.object({
			email: z.string().email(),
		})
		.optional()
		.nullable(),
	is_shared: z.boolean().nullish(),
	deleted: z.boolean().nullish(),
	filters: z.record(z.any()).nullish(),
	variables: z.record(z.any()).nullish(),
	tags: z.array(z.string()).nullish(),
	tiles: z.array(z.record(z.any())).nullish(),
});

// Input schema for creating dashboards
export const CreateDashboardInputSchema = z.object({
	name: z.string().min(1, "Dashboard name is required"),
	description: z.string().optional(),
	pinned: z.boolean().optional().default(false),
	tags: z.array(z.string()).optional(),
});

// Input schema for updating dashboards
export const UpdateDashboardInputSchema = z.object({
	name: z.string().optional(),
	description: z.string().optional(),
	pinned: z.boolean().optional(),
	tags: z.array(z.string()).optional(),
});

// Input schema for listing dashboards
export const ListDashboardsSchema = z.object({
	limit: z.number().int().positive().optional(),
	offset: z.number().int().nonnegative().optional(),
	search: z.string().optional(),
	pinned: z.boolean().optional(),
});

// Input schema for adding insight to dashboard
export const AddInsightToDashboardSchema = z.object({
	insightId: z.number().int().positive(),
	dashboardId: z.number().int().positive(),
});

// Type exports
export type PostHogDashboard = z.infer<typeof DashboardSchema>;
export type CreateDashboardInput = z.infer<typeof CreateDashboardInputSchema>;
export type UpdateDashboardInput = z.infer<typeof UpdateDashboardInputSchema>;
export type ListDashboardsData = z.infer<typeof ListDashboardsSchema>;
export type AddInsightToDashboardInput = z.infer<typeof AddInsightToDashboardSchema>;
